File: Euler034.java
public class Euler034
{
    public static int[] FACTS = { 1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880 };

    public static void main (String[] args) {
        int sum = 0;
        for (int n = 10; n < FACTS[9]*7; n++) {
            int tmp = n, dsum = 0;
            do {
                dsum += FACTS[tmp%10];
                tmp /= 10;
            } while (tmp > 0);
            if (n == dsum) {
                sum += n;
            }
        }
        System.out.println(sum);
    }
}

File: Euler034.scala
object Euler034s extends EulerApp {
  val facts = Array(1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880)
  def digfact (sum :Int, n :Int) :Int =
    if (n == 0) sum else digfact(facts(n % 10) + sum, n/10)
  def check (sum :Int, n :Int) :Int =
    if (n == 2000000) sum
    else if (digfact(0, n) == n) check(sum+n, n+1)
    else check(sum, n+1)
  def answer = check(0, 10)
}
